### ISSUE TITLE
how to add bias to to attn

### ISSUE THREAD

**Issue by maidabu** [Dao-AILab/flash-attention#534]
If original code is:
  ...
  attn = q @ k
  attn = attn.softmax(dim=-1)
  ...

now my code is:
  ...
  attn = q @ k
  attn = attn + bias
  attn = attn.softmax(dim=-1)
  ...
where the shape of bias is the same with attn, and the dtype of bias is float(or half).

How do I modify the compute_attn_1rowblock in the flash_fwd_kernel.h.
I tried to copy global bias to the acc_s, but never successfully.
Any one can help me.

**Comment by tridao** [comment]
acc_s is stored in a permuted layout. E.g. thread 0 holds element at indices (0, 0), (0, 1), (0, 8), (0, 9), etc, since this is the layout of the accumulator returned by the mma instruction (tensor cores). 
The layout is documented by Nvidia [here](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#matrix-fragments-for-mma-m16n8k16-with-floating-point-type).

So you'd need to read the bias in such a way that each thread would have the bias at the indices corresponding to the indices of their acc_s.
The right way to do that is probably to use an intermediate buffer in shared memory: read from global memory to shared memory, then use the `ldmatrix` instruction to load from shared memory to registers of each thread. Cutlass has abstractions for these (we're using them to read in Q, K, V) but it'll take some trial and error.

**Comment by maidabu** [comment]
> 

Thanks very much for replying.
I found that you copy result of softmax to global memory directly without usage of shared memory in write_softmax_to_gmem. Why not use shared memory here?
I tried to copy global bias to acc_s referred to the write_softmax_to_gmem in flash_fwd_kernel.h, but got incorrect result. Is the idea wrong?

**Comment by tridao** [comment]
The softmax is written to gmem in a permuted layout, for simplicity. This softmax is only used for testing. So when we test, we "unpermute" this matrix using Pytorch [here](https://github.com/Dao-AILab/flash-attention/blob/08c295c043ed9463f0f3d19de67e03f469455fd2/tests/test_flash_attn.py#L327). This is slow, but it doesn't matter because it's only for testing.

**Comment by maidabu** [comment]
> The softmax is written to gmem in a permuted layout, for simplicity. This softmax is only used for testing. So when we test, we "unpermute" this matrix using Pytorch [here](https://github.com/Dao-AILab/flash-attention/blob/08c295c043ed9463f0f3d19de67e03f469455fd2/tests/test_flash_attn.py#L327). This is slow, but it doesn't matter because it's only for testing.

@tridao 

I see. I tried to load global bias to shared memory, and then to registers of each thread. But not successfully. Is there any sample for reference.

**Comment by tridao** [comment]
You can look at how Q, K, V are loaded, or how O is written out. The layouts are configured in `kernel_traits.h` to match the MMA layout. The most relevant is probably the O layout (since acc_o is also the output of an mma instruction, just like acc_s).

**Comment by tridao** [comment]
Tbh it involves some trial and error to get the layout right. You can use the print function to print out the values (see some commented out code in `flash_fwd_kernel.h` to see if what you're loading in matches what you expect.

**Comment by maidabu** [comment]
> Tbh it involves some trial and error to get the layout right. You can use the print function (`cute::print_tensor`) to print out the values (see some commented out code in `flash_fwd_kernel.h`) to see if what you're loading in matches what you expect.

@tridao 

Shape of Q, K, acc_o is MxK, NxK, MxK, but shape of bias is MxN which is the same with Q@K^T. Does flash:copy support copy global bias to shared memory?

**Comment by maidabu** [comment]
> That's why you'll have to play with the shape and layout to get it right. flash::copy should work (it's just calling cute::copy under the hood).

@tridao 
Thank you.
Another question, after copying bias to acc_s, and then flash::gemm, the data in acc_s is the same with the pytorch result(attn = q @ k^T + bias) in theory. Is it?
